decisions/websre/0004-use-1-kubernetes-cluster-per-environment.md

```diff
@@ -0,0 +1,56 @@
+1. Having a distinct Kubernetes cluster per environment;
+2. Having one namespace per environment in a shared Kubernetes cluster;
+3. Having distinct node pool(s) per environment within a single cluster along with using selectors, taints, and tolerations.
+
+## Decision
+
+For Web SRE Services persistent environments - dev, stage, prod - that are Kubernetes-based, we will:
+* Use a distinct Kubernetes cluster per environment (dev, stage, prod) within the same infrastructure account;
+* Use a distinct Kubernetes namespace within each cluster per application;
+* Deploy applications to these distinct clusters as relevant to the expected application environment (e.g. applications should deploy to the stage cluster, prod applications should deploy to the prod cluster, etc.);
+* Deploy operationals & infrastructure changes & upgrades to these distinct clusters, specifically moving changes from stage to prod within an appropriate timeframe, with an end goal of keeping stage and prod clusters in close alignment;
+* Development clusters availability and uptime expectations (of the application or the infrastructure) does not include support outside of business hours;
+* Development clusters can optionally include automatic infrastructure or platform updates, as a first place to review these changes and their effects;
+* Only use reviewed & approved, production-ready stable application & infrastructure releases on the Production cluster;
+* Only use reviewed & approved, stable infrastructure releases on the Stage cluster, and deploy those infrastructure changes when able to relatively quickly move the changes from stage into prod to avoid divergent results from staging and production applications.
+
+This ADR does not include the following:
+* How the applications within each cluster name their namespaces (e.g. keeping "discourse_prod" namespace in a production cluster is redundent, but also very explicit, and is left up to the SREs working on the service);
+* An immediate push to migrate all existing services to this pattern, as it will cause a lot of churn. Instead, this should be used going forward when doing migrations or starting new services;
+* Explicit declarations on SLOs, SLAs, or uptime guarantees on these environment clusters beyond the declaration that Development clusters do not have round-the-clock uptime guarantees to make space for more frequent platform and application changes there;
+* Any expectation on multi-tenant clusters versus single-tenant clusters; the pattern for environments should apply in either case;
+* An expectation that a development cluster exists - this is left up to a service's need(s) and SREs' determination;
+* Decisions on approaches to non-peristent clusters, e.g. Testing environments;
+* Decisions on how to implement 1 cluster per environment, e.g. it can follow the itse-apps-*-1 implementation or some other approach as the SRE sees fit.
```

Comment:
Something else not mentioned is any required sequencing. E.g. can an app be deployed to prod without first having been deployed to stage? (The old cloudops team has this, and speaks of "promoting" an app from stage to prod to reinforce.)

I suspect that not all of your apps warrant a mandatory soak in stage, so wording may be difficult.

Reply:
I left out that we (secops) have tooling we can hook up to staging environments to run sanity checks before promoting to prod. If you're interested, ping Greg.

Reply:
So, specific application release & deploy sequencing isn't necessarily the focus here, since lots of projects in the web sre portfolio already have their specific approaches & solutions to this already (usually inherited from whatever team owned the infrastructure pre-web sre merge). 

The only place in this particular proposal where sequencing comes up is we do want to keep stage infrastructure & services aligned with production infrastructure & services. I can try to add something more specific there simply focused on the applications' themselves.

Reply:
depending on what sanity checks you're thinking about @hwine & @g-k , this could be a useful addition. We do have quite a few services we manage already that have pre-production integration & e2e tests as a baseline, and lot of the more advance services have more.

Reply:
We run a ZAP scan after the stage deploy and before the prod deploy for some of the cloudops-deployment pipelines e.g. https://github.com/mozilla-services/cloudops-deployment/blob/master/projects/autograph/pipeline-edge.groovy#L56-L61

This might a nice option in k8s, but it does add latency to the deploy pipeline and I don't have data on whether the step has found or prevented any issues in the cloudops prod env.

Reply:
Sorry, accidentally resolved.

Can note ZAP scanning as a possible add on & compare if we have alternatives in place in our JIRA ticket. iirc, pastebin & etherpad both have docker image security scanning as remediation steps, so fits in there.

Doesn't really fit into this ADR though, given its specifically about infrastructure platform per environment / tier, and we don't want to go further in scope than that here to keep the flexibility we need for all the infrastructure & processes we currently manage. Expect something specific like application release & deploy pipelines to come up in a future ADR when we're maybe a little further along in consolidation all the different standards & approaches we're handling atm within web sre.